shell/platform/linux/public/flutter_linux/fl_event_channel.h

```diff
@@ -0,0 +1,187 @@
+ *
+ * The following example shows how to send events on a channel:
+ *
+ * |[<!-- language="C" -->
+ * static FlEventChannel *channel = NULL;
+ * static gboolean send_events = FALSE;
+ *
+ * static void event_occurs_cb (FooEvent *event) {
+ *   if (send_events) {
+ *     g_autoptr(FlValue) message = foo_event_to_value (event);
+ *     g_autoptr(GError) error = NULL;
+ *     if (!fl_event_channel_send (channel, message, NULL, &error)) {
+ *       g_warning ("Failed to send event: %s", error->message);
+ *     }
+ *   }
+ * }
+ *
+ * static FlMethodErrorResponse* listen_cb (FlEventChannel* channel,
+ *                                          FlValue *args,
+ *                                          gpointer user_data) {
+ *   send_events = TRUE;
+ *   return NULL;
+ * }
+ *
+ * static FlMethodErrorResponse* cancel_cb (GObject *object,
+ *                                          FlValue *args,
+ *                                          gpointer user_data) {
+ *   send_events = FALSE;
+ *   return NULL;
+ * }
+ *
+ * static void setup_channel () {
+ *   g_autoptr(FlStandardMethodCodec) codec = fl_standard_method_codec_new ();
+ *   channel = fl_event_channel_new (messenger, "flutter/foo",
+ *                                   FL_METHOD_CODEC (codec),
+ *                                   listen_cb, cancel_cb, NULL, NULL);
+ * }
+ * ]|
+ *
+ * #FlEventChannel matches the EventChannel class in the Flutter
+ * services library.
+ */
+
+/**
+ * FlEventChannelResponseHandle:
+ *
+ * #FlEventChannelResponseHandle is an object used to send responses
+ * with.
+ */
+
+/**
+ * FlEventChannelHandler:
+ * @channel: an #FlEventChannel.
+ * @args: arguments passed from the Dart end of the channel.
+ * @user_data: (closure): data provided when registering this handler.
+ *
+ * Function called when the stream is listened to or cancelled.
+ *
+ * Returns: (transfer full): an #FlMethodErrorResponse or %NULL if no error.
+ */
+typedef FlMethodErrorResponse* (*FlEventChannelHandler)(FlEventChannel* channel,
+                                                        FlValue* args,
+                                                        gpointer user_data);
+
+/**
+ * fl_event_channel_new:
+ * @messenger: an #FlBinaryMessenger.
+ * @name: a channel name.
+ * @codec: the message codec.
+ * @listen_handler: (allow-none): function to call when the Dart side of the
+ * channel starts listening to the stream.
+ * @cancel_handler: (allow-none): function to call when the Dart side of the
+ * channel cancels their subscription to the stream.
+ * @user_data: (closure): user data to pass to @listen_handler and
+ * @cancel_handler.
+ * @destroy_notify: (allow-none): a function which gets called to free
+ * @user_data, or %NULL.
+ *
+ * Creates an event channel. @codec must match the codec used on the Dart
+ * end of the channel.
+ *
+ * Returns: a new #FlEventChannel.
+ */
+FlEventChannel* fl_event_channel_new(FlBinaryMessenger* messenger,
```

Comment:
My thinking at the time was it was more complex to have them set separately due to having to support replacing an existing handler. It also seemed more risky to allow the API user to change these as they're essential for correct stream behaviour.

Looking at your examples it does seem cleaner and more consistent with the `FlMethodChannel` so I've changed them. I called the method `fl_event_channel_set_stream_handlers` to avoid the double *event* in the name. The method name still seems a bit vague/confusing, but I don't know if there's a better option.

Thanks for the review!

Reply:
fl_event_channel_set_stream_handlers() sounds better :+1: